### PR TITLE
feat(trading): datanode block header for env and node switcher

### LINKED ISSUE #1791: Update nodeHealth to use `x-block-height` header to check data node health
#2701 improved the way `NodeHealth` works, but misses one imporant part of the ticket. Data Node adds some metadata headers on every request:

```
< x-block-height: 926035
< x-block-timestamp: 1675424066275232000
< x-vega-connection: CONNECTED
< x-vega-node-id: n11.testnet.vega.xyz
```

These are interesting because they contain the data node's idea of the block height, timestamp, and whether or not it is connected to a core vega node. This is different to the `blockHeight` property in statistics, which comes directly from the  core node.

To put it another way, if you run `curl -v https://api.n11.testnet.vega.xyz/statistics`, the response header `x-block-height` could contain a different height to the body property `.statistics.blockHeight`. It this is the case, it is an indicator that the data you are being served is not up to date, which is a very important thing to flag up to the user.

Screenshot 2023-02-03 at 12 08 41

## Acceptance Criteria

- [ ] As a user I can clearly see that the app I am on is connected to a datanode that isn’t up to date 

## Tasks

- [ ] Update `Nodehealth` component to log `x-block-height` for the data node you are connected to
- [ ] For the data node you are connected to, compare `x-block-height` to the currently held block height value (i.e. the one read from statistics)
- [ ] Use the currently acceptable value of `BLOCK_THRESHOLD` - if `x-block-height` is more than `BLOCK_THRESHOLD` *behind* `statistics.blockHeight`, show an error indicating stale data

### PR DESCRIPTION
**Pull request by mattrussell36**
# Related issues 🔗

Closes #1791 

# Description ℹ️

- Changes node health UI in trading footer to be determined by comparing header block height from the data node, and the 'core' block height in the statistics query. This has the benefit of not requiring multiple requests across different nodes to determine overall node health.
- NodeSwitcher component has also been updated to use datanode block height rather than the query result block height.
- Store selected node in localStorage and attempts to use that on startup

# Technical 👨‍🔧

- Adds a custom ApolloLink store a map of datanode urls to header values
- Environment hook and NodeSwitcher dialog have been refactored so that additional nodes state and client instance is only created in the dialog itself and garbage collected when the dialog is closed.
- Adds NodeGuard component to trading app to prevent rendering app connected to a non functioning node
- Adds AppFailure with DynamicLoader component fallback/loading states consistent

### CODE REVIEW

**Comment by daunatv** [comment]
Did a CMD+SHIFT+R, still same, created a new incognito tab, still offline, accessed using Firefox and it was online.
https://user-images.githubusercontent.com/100769129/218757546-2139ebb7-ae43-4392-af44-60759f8bfa57.mov

File: apps/liquidity-provision-dashboard/src/app/app.tsx [hunk 1]
```diff
@@ -1,18 +1,46 @@
+import type { InMemoryCacheConfig } from '@apollo/client';
+import { NetworkLoader, useInitializeEnv } from '@vegaprotocol/environment';
 import { useRoutes } from 'react-router-dom';
 
 import '../styles.scss';
 import { Navbar } from './components/navbar';
 
 import { routerConfig } from './routes/router-config';
 
+const cache: InMemoryCacheConfig = {
```
Comment (edd) [inline apps/liquidity-provision-dashboard/src/app/app.tsx:10]: **thought (non-blocking)**: @asiaznik added a shared config in https://github.com/vegaprotocol/frontend-monorepo/pull/2914, which might be useful. Or they might be totally separate. Worth a quick look anyway - perhaps there's some benefit to all apps ffrom these cache changes.

File: libs/environment/src/hooks/use-environment.spec.ts [hunk 2]
```diff
@@ -0,0 +1,259 @@
...
+};
+
+describe('useEnvironment', () => {
+  const env = process.env;
+
+  beforeEach(() => {
+    jest.resetModules();
+    process.env = { ...env };
+
+    process.env['NX_VEGA_ENV'] = mockEnvVars.VEGA_ENV;
```
Comment (edd) [inline libs/environment/src/hooks/use-environment.spec.ts:64]: **thought (non-blocking)**: Seeing this made me realise what a big task removing our dependence on env vars is going to be. But staying consistent here is definitely the right thing for now.

**Review by edd — APPROVED** [review]
👍 Two non blocking thoughts occured to me while looking at this. Good test coverage, looking forward to seeing it in

**Review by notbot00 — APPROVED** [review]

File: apps/liquidity-provision-dashboard/src/app/app.tsx (hunk 1 above)
Comment (mattrussell36) [reply apps/liquidity-provision-dashboard/src/app/app.tsx:10]: This config looks different, to the new default one. I'll leave this as is for now. Don't wan't to introduce anything unexpected

File: libs/environment/src/components/node-guard/node-guard.tsx [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+  children,
+  failure,
+  skeleton,
+}: {
+  children: ReactNode;
+  failure: ReactNode;
+  skeleton: ReactNode;
+}) => {
+  const { error, loading } = useStatisticsQuery();
+  const nonIdealWrapperClasses =
```
[the remark this answers is not in the document]
Comment (mattrussell36) [reply libs/environment/src/components/node-guard/node-guard.tsx]: Renamed to be a bit more generic

**Pull request merged**